File: ios/Sources/HealthBridgeKit/Pairing.swift
// Swift mirror of cli/internal/pairing/pairing.go for the iOS responder.
//
// Roles:
//   - The CLI is the *initiator*. It mints the pair_id, generates an
//     X25519 keypair, posts the pubkey, and renders the resulting
//     PairLink (`{v, pair_id, cli_pub_hex, relay_url}`) as a QR.
//   - The iOS app is the *responder*. It scans the QR, generates its own
//     X25519 keypair, posts the pubkey, derives the same session key, and
//     shows the SAS for the user to confirm against what the CLI shows.
//
// This file is the iOS half. It is intentionally pure: no UIKit, no
// HealthKit, no Keychain. The pairing UI imports this and persists the
// returned PairResult through whatever storage layer it owns.

import Foundation
import CryptoKit

/// PairLink mirrors `pairing.PairLink` in Go. The CLI prints one of these
/// as a JSON-encoded QR code; the iOS app decodes it after a successful
/// scan and feeds it to `respondPairing`.
public struct PairLink: Codable, Sendable, Equatable {
    public let pairID: String
    public let cliPubHex: String
    public let relayURL: String
    public let version: String

    enum CodingKeys: String, CodingKey {
        case pairID = "pair_id"
        case cliPubHex = "cli_pub_hex"
        case relayURL = "relay_url"
        case version = "v"
    }

    public init(pairID: String, cliPubHex: String, relayURL: String, version: String = HealthBridgePairing.protocolVersion) {
        self.pairID = pairID
        self.cliPubHex = cliPubHex
        self.relayURL = relayURL
        self.version = version
    }
}

/// What the iOS responder gets back from a successful pairing.
public struct PairResult: Sendable, Equatable {
    public let pairID: String
    public let relayURL: String
    public let sessionKey: Data
    public let authToken: String
    public let iosPub: Data
    public let iosPriv: Data
    public let cliPub: Data
    public let sas: String

    public init(
        pairID: String,
        relayURL: String,
        sessionKey: Data,
        authToken: String,
        iosPub: Data,
        iosPriv: Data,
        cliPub: Data,
        sas: String
    ) {
        self.pairID = pairID
        self.relayURL = relayURL
        self.sessionKey = sessionKey
        self.authToken = authToken
        self.iosPub = iosPub
        self.iosPriv = iosPriv
        self.cliPub = cliPub
        self.sas = sas
    }
}

/// Errors raised by the pairing helpers. Each case maps to a clear UI
/// message in the iOS pairing screen.
public enum PairingError: Error, Equatable {
    case invalidLinkJSON
    case unsupportedProtocolVersion(String)
    case missingLinkFields
    case badCLIPubHex
    case wrongPubkeySize
    case relayMissingAuthToken
}

public enum HealthBridgePairing {

    /// Wire-format version. Must match `pairing.protocolVersion` in Go.
    /// v2 is the QR-from-CLI flow; v1 was the (now-removed) QR-from-iOS
    /// flow with `ios_pub_hex` instead of `cli_pub_hex`.
    public static let protocolVersion = "healthbridge-pair-v2"

    /// Decode a PairLink from the JSON string the CLI encoded into the
    /// QR code. Validates protocol version *first* so an old v1 link
    /// (which is missing the v2 field shape entirely) surfaces as
    /// `unsupportedProtocolVersion` rather than `invalidLinkJSON`.
    public static func decodeLink(_ json: String) throws -> PairLink {
        guard let data = json.data(using: .utf8) else {
            throw PairingError.invalidLinkJSON
        }
        // Parse loosely first so the version check runs even if v2-shaped
        // fields are missing.
        guard let raw = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw PairingError.invalidLinkJSON
        }
        let version = raw["v"] as? String ?? ""
        if version != protocolVersion {
            throw PairingError.unsupportedProtocolVersion(version)
        }
        let pairID = raw["pair_id"] as? String ?? ""
        let cliPubHex = raw["cli_pub_hex"] as? String ?? ""
        let relayURL = raw["relay_url"] as? String ?? ""
        if pairID.isEmpty || cliPubHex.isEmpty || relayURL.isEmpty {
            throw PairingError.missingLinkFields
        }
        guard let pub = Data(hexString: cliPubHex) else {
            throw PairingError.badCLIPubHex
        }
        if pub.count != HealthBridgeCrypto.publicKeySize {
            throw PairingError.wrongPubkeySize
        }
        return PairLink(pairID: pairID, cliPubHex: cliPubHex, relayURL: relayURL, version: version)
    }

    /// Run the iOS (responder) half of the protocol. Generates a fresh
    /// X25519 keypair, posts the pubkey to the relay, derives the shared
    /// session key, and returns a PairResult.
    ///
    /// The caller is responsible for displaying the SAS, asking the user
    /// to confirm it matches the CLI's, and persisting the result.
    public static func respondPairing(client: RelayClient, link: PairLink) async throws -> PairResult {
        let cliPub = try requireHex(link.cliPubHex)
        if cliPub.count != HealthBridgeCrypto.publicKeySize {
            throw PairingError.wrongPubkeySize
        }
        let keys = HealthBridgeCrypto.KeyPair()
        let iosPub = keys.publicKeyBytes
        let iosPriv = keys.privateKey.rawRepresentation

        let state = try await client.postPubkey(
            side: "ios",
            pubkeyHex: iosPub.hexString
        )
        guard let token = state.authToken, !token.isEmpty else {
            throw PairingError.relayMissingAuthToken
        }

        let shared = try HealthBridgeCrypto.sharedSecret(myPriv: iosPriv, theirPub: cliPub)
        let transcript = HealthBridgeCrypto.buildTranscript(pairID: link.pairID, iosPub: iosPub, cliPub: cliPub)
        let sessionSym = HealthBridgeCrypto.deriveSessionKey(
            shared: shared,
            transcript: transcript,
            info: "healthbridge/m2/session"
        )
        let sessionData = sessionSym.withUnsafeBytes { Data($0) }
        let sas = HealthBridgeCrypto.sas(shared: shared, transcript: transcript)

        return PairResult(
            pairID: link.pairID,
            relayURL: link.relayURL,
            sessionKey: sessionData,
            authToken: token,
            iosPub: iosPub,
            iosPriv: iosPriv,
            cliPub: cliPub,
            sas: sas
        )
    }

    private static func requireHex(_ s: String) throws -> Data {
        guard let d = Data(hexString: s) else {
            throw PairingError.badCLIPubHex
        }
        return d
    }
}

File: cli/internal/pairing/pairing.go
// Package pairing implements the M2 X25519 + relay /pair flow.
//
// Roles in the new flow:
//
//   - The CLI is the *initiator*. `healthbridge pair` mints a pair_id,
//     generates an X25519 keypair, posts its pubkey to the relay, and
//     prints a QR encoding `{pair_id, cli_pub_hex, relay_url}` to the
//     terminal. It then long-polls the relay for the iOS side.
//
//   - The iOS app is the *responder*. The user opens HealthBridge → Pair,
//     scans the QR with the camera, the app generates its own X25519
//     keypair, posts its pubkey to the relay, derives the shared session
//     key, and shows a 6-digit SAS that the user confirms matches what
//     the CLI is showing.
//
// Both sides end up with the same Result. The relay only ever sees the
// two pubkeys and an opaque per-pair auth_token; it cannot read any
// later jobs or results, which are sealed under the derived session key.
//
// The two helpers in this file (InitiatePairing and RespondPairing) are
// pure functions over a relay.Client; persistence is the caller's job.
// HealthBridgeKit/Pairing.swift is a byte-identical Swift mirror.
package pairing

import (
	"context"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"

	"github.com/shuyangli/healthbridge/cli/internal/crypto"
	"github.com/shuyangli/healthbridge/cli/internal/relay"
)

// PairLink is the JSON the CLI encodes into a QR code at the start of
// pairing. The iOS app scans one of these and uses it to drive
// RespondPairing.
type PairLink struct {
	PairID   string `json:"pair_id"`
	CLIPub   string `json:"cli_pub_hex"`
	RelayURL string `json:"relay_url"`
	// Version is bumped when the pairing protocol changes. Mismatched
	// versions abort pairing rather than silently produce wrong session
	// keys.
	Version string `json:"v"`
}

// protocolVersion identifies the current pairing wire format. This was
// bumped from v1 to v2 when the initiator role moved from iOS to CLI;
// the field shape changed (`ios_pub_hex` → `cli_pub_hex`) so old links
// must fail loudly rather than silently produce a wrong session key.
const protocolVersion = "healthbridge-pair-v2"

// EncodeLink renders a PairLink as a compact JSON string suitable for a
// QR code or pasted token.
func EncodeLink(link *PairLink) (string, error) {
	if link.Version == "" {
		link.Version = protocolVersion
	}
	b, err := json.Marshal(link)
	if err != nil {
		return "", fmt.Errorf("pairing: marshal link: %w", err)
	}
	return string(b), nil
}

// DecodeLink parses a JSON string back into a PairLink and validates the
// version.
func DecodeLink(s string) (*PairLink, error) {
	var link PairLink
	if err := json.Unmarshal([]byte(s), &link); err != nil {
		return nil, fmt.Errorf("pairing: parse link: %w", err)
	}
	if link.Version != protocolVersion {
		return nil, fmt.Errorf("pairing: unsupported protocol version %q", link.Version)
	}
	if link.PairID == "" || link.CLIPub == "" || link.RelayURL == "" {
		return nil, errors.New("pairing: link missing required fields")
	}
	if _, err := hex.DecodeString(link.CLIPub); err != nil {
		return nil, fmt.Errorf("pairing: cli_pub not hex: %w", err)
	}
	return &link, nil
}

// Result is what either side ends up with after a successful pairing.
type Result struct {
	PairID     string
	RelayURL   string
	SessionKey []byte
	AuthToken  string
	IOSPub     []byte
	CLIPub     []byte
	// Private key of the local side. CLI side fills CLIPriv; iOS side
	// fills IOSPriv. The other field is left nil. We persist the
	// initiator's private key only because it must survive across the
	// long-poll wait for the responder.
	CLIPriv []byte
	IOSPriv []byte
	SAS     string
}

// InitiatePairing runs the CLI (initiator) half of the protocol. It
// generates a fresh X25519 keypair, commits the pubkey to the relay, and
// returns a PairLink the caller should encode into a QR for the iOS app
// to scan. The Result returned at this stage has CLIPriv populated but
// no SessionKey/SAS — those come from CompletePairing once the iOS side
// has responded.
func InitiatePairing(ctx context.Context, c *relay.Client, relayURL string) (*Result, *PairLink, error) {
	if c.PairID == "" {
		return nil, nil, errors.New("pairing: client PairID not set")
	}
	keys, err := crypto.GenerateKeyPair()
	if err != nil {
		return nil, nil, fmt.Errorf("pairing: gen cli keypair: %w", err)
	}
	if _, err := c.PostPubkey(ctx, "cli", hex.EncodeToString(keys.Public)); err != nil {
		return nil, nil, fmt.Errorf("pairing: post cli pubkey: %w", err)
	}
	r := &Result{
		PairID:   c.PairID,
		RelayURL: relayURL,
		CLIPub:   keys.Public,
		CLIPriv:  keys.Private,
	}
	link := &PairLink{
		PairID:   c.PairID,
		CLIPub:   hex.EncodeToString(keys.Public),
		RelayURL: relayURL,
		Version:  protocolVersion,
	}
	return r, link, nil
}

// CompletePairing finishes the CLI (initiator) side after the iOS app
// has scanned the QR and posted its pubkey. It long-polls the relay for
// the iOS pubkey + auth_token, derives the session key, and fills in
// SessionKey/AuthToken/IOSPub/SAS on the Result returned by InitiatePairing.
func CompletePairing(ctx context.Context, c *relay.Client, partial *Result, waitMs int) (*Result, error) {
	if partial == nil || len(partial.CLIPriv) == 0 || len(partial.CLIPub) == 0 {
		return nil, errors.New("pairing: CompletePairing requires a partial Result from InitiatePairing")
	}
	state, err := c.PollPair(ctx, waitMs)
	if err != nil {
		return nil, fmt.Errorf("pairing: poll pair: %w", err)
	}
	if state.IOSPub == nil || state.AuthToken == nil {
		return nil, errors.New("pairing: timed out waiting for ios pubkey")
	}
	iosPub, err := hex.DecodeString(*state.IOSPub)
	if err != nil {
		return nil, fmt.Errorf("pairing: bad ios pubkey: %w", err)
	}
	if len(iosPub) != crypto.PublicKeySize {
		return nil, fmt.Errorf("pairing: ios pubkey size %d, want %d", len(iosPub), crypto.PublicKeySize)
	}
	shared, err := crypto.SharedSecret(partial.CLIPriv, iosPub)
	if err != nil {
		return nil, fmt.Errorf("pairing: shared secret: %w", err)
	}
	transcript := crypto.BuildTranscript(partial.PairID, iosPub, partial.CLIPub)
	session, err := crypto.DeriveSessionKey(shared, transcript, "healthbridge/m2/session")
	if err != nil {
		return nil, fmt.Errorf("pairing: derive session: %w", err)
	}
	out := *partial
	out.IOSPub = iosPub
	out.SessionKey = session
	out.AuthToken = *state.AuthToken
	out.SAS = crypto.SAS(shared, transcript)
	return &out, nil
}

// RespondPairing runs the iOS (responder) half of the protocol given a
// PairLink decoded from the QR the CLI showed. It generates an X25519
// keypair, commits the pubkey to the relay, derives the session key,
// and returns a Result the caller should display (SAS) and persist.
//
// This helper exists in Go so end-to-end tests can drive the iOS side
// from the same process as the CLI side. The production iOS app has its
// own Swift mirror in HealthBridgeKit/Pairing.swift.
func RespondPairing(ctx context.Context, c *relay.Client, link *PairLink) (*Result, error) {
	if c.PairID != link.PairID {
		return nil, fmt.Errorf("pairing: client pair_id %q != link pair_id %q", c.PairID, link.PairID)
	}
	cliPub, err := hex.DecodeString(link.CLIPub)
	if err != nil {
		return nil, fmt.Errorf("pairing: bad cli_pub: %w", err)
	}
	if len(cliPub) != crypto.PublicKeySize {
		return nil, fmt.Errorf("pairing: cli_pub size %d, want %d", len(cliPub), crypto.PublicKeySize)
	}
	keys, err := crypto.GenerateKeyPair()
	if err != nil {
		return nil, fmt.Errorf("pairing: gen ios keypair: %w", err)
	}
	state, err := c.PostPubkey(ctx, "ios", hex.EncodeToString(keys.Public))
	if err != nil {
		return nil, fmt.Errorf("pairing: post ios pubkey: %w", err)
	}
	if state.AuthToken == nil {
		return nil, errors.New("pairing: relay did not issue auth_token (cli side missing?)")
	}
	shared, err := crypto.SharedSecret(keys.Private, cliPub)
	if err != nil {
		return nil, err
	}
	transcript := crypto.BuildTranscript(link.PairID, keys.Public, cliPub)
	session, err := crypto.DeriveSessionKey(shared, transcript, "healthbridge/m2/session")
	if err != nil {
		return nil, err
	}
	return &Result{
		PairID:     link.PairID,
		RelayURL:   link.RelayURL,
		SessionKey: session,
		AuthToken:  *state.AuthToken,
		IOSPub:     keys.Public,
		IOSPriv:    keys.Private,
		CLIPub:     cliPub,
		SAS:        crypto.SAS(shared, transcript),
	}, nil
}
